setup/helpers/abs_helper.rb

```diff
@@ -0,0 +1,657 @@
+        if !valid_abs_response?(response)
+          invalid_response = true
+          puts "Unable to provision host for role: #{role}"
+          puts
+
+          if hosts.empty?
+            puts "No ABS hosts were provisioned"
+            puts
+          else
+            puts "Returning any provisioned hosts"
+            return_abs_resource_hosts(hosts.to_json)
+          end
+
+          # stop requesting hosts
+          break
+
+        else
+          # reformat to satisfy beaker
+          hosts << JSON.parse(reformat_abs_resource_host(response.body))
+        end
+      end
+
+      if !hosts.empty? && !invalid_response
+        abs_resource_hosts = hosts.to_json
+        ENV["ABS_RESOURCE_HOSTS"] = abs_resource_hosts
+        puts "ABS_RESOURCE_HOSTS=#{ENV['ABS_RESOURCE_HOSTS']}"
+
+        # write to "last_abs_resource_hosts.log" (used when returning hosts)
+        update_last_abs_resource_hosts(abs_resource_hosts)
+      end
+
+    else
+      puts "Unable to proceed without a valid ABS token"
+      puts
+    end
+
+    # only return the hosts if they're successfully verified
+    success = verify_abs_hosts(hosts) unless abs_resource_hosts.nil?
+    verified_hosts = success ? abs_resource_hosts : nil
+
+    if verified_hosts
+      puts
+      puts "ABS hosts have been successfully provisioned"
+      puts "Pausing to ensure successful configuration"
+
+      # TODO: remove?
+      sleep 30
+    end
+
+    return verified_hosts
+  end
+
+  # Returns the specified ABS hosts via ABS
+  #
+  # @author Bill Claytor
+  #
+  # @param [JSON] abs_resource_hosts The hosts to de-provision
+  #
+  # @return [JSON] The hosts that were successfully returned
+  #
+  # @example
+  #   returned_hosts = abs_return_resource_hosts(abs_resource_hosts)
+  #
+  def return_abs_resource_hosts(abs_resource_hosts)
+    returned_hosts = nil
+    puts "ABS hosts specified for return: #{abs_resource_hosts}"
+
+    is_valid = valid_abs_resource_hosts?(abs_resource_hosts)
+    unless is_valid
+      puts "De-provisioning via awsdirectreturn requires an array of hostnames to be specified"
+      puts "Specify hostnames via the ABS_RESOURCE_HOSTS environment variable"
+      puts "Or specify via the last_abs_resource_hosts.log file"
+      puts
+    end
+
+    has_token = abs_initialize
+    unless has_token
+      puts "Unable to proceed without a valid ABS token"
+      puts
+    end
+
+    if is_valid && has_token
+      uri = URI("#{@abs_base_url}/awsdirectreturn")
+
+      # returned_hosts = []
+      hosts = JSON.parse(abs_resource_hosts)
+      hosts.each do |host|
+        hostname = host["hostname"]
+
+        puts "Returning host: #{hostname}"
+        body = get_awsdirectreturn_request_body(hostname)
+        res = perform_awsdirect_request(uri, body)
+
+        if valid_abs_response?(res)
+          puts "Successfully returned host: #{hostname}"
+          returned_hosts ? returned_hosts << host : returned_hosts = [host]
```

Comment:
If you initialize 'returned_hosts = []' instead of nil you can avoid this logic and just use 'returned_hosts << host'